Assistant: add preview support for snowflake cortex provider

sharon-wang opened the pull request:
### Summary

- addresses https://github.com/posit-dev/positron/issues/8613
- addresses https://github.com/posit-dev/positron/issues/10748

#### Auth Methods

- API Key (Snowflake PAT) + Account Identifier/Snowflake Base Url
- Workbench Managed Credentials (auto-detect OAuth tokens)

#### Models

- Hardcoded list that can be overridden with `positron.assistant.configuredModels` using the `snowflake-cortex` provider ID
    - Claude Sonnet 4
    - Claude Sonnet 4.5
    - Claude Haiku 4.5
    - GPT-5
    - GPT-4.1

#### OpenAI API integration

- Build on top of our existing OpenAI and OpenAI-Compatible handling with the Vercel AI SDK v4, with some additional Snowflake-specific handling (auth and errors)
    - Does not use OpenAI Node.js SDK nor the Snowflake REST API
    - Includes several workarounds for AI SDK v4 compatibility, including parameter transformations (`max_tokens` --> `max_completion_tokens`; removal of `temperature`). We will be able to remove these workarounds if/when we upgrade to AI SDK v5
- Snowflake-specific handling
    - Extra error message handling for cross-region inference issues and network policy authentication failures
    - Some Snowflake responses are returned without roles (empty string returned as role) -- these are updated with role assistant to avoid errors

#### Some screenshots

##### Before logging in with API key
<img width="605" height="455" alt="image" src="https://github.com/user-attachments/assets/5658f220-6091-49e6-96e2-77b3132f265a" />

##### After Managed Creds auto-login
<img width="605" height="455" alt="image" src="https://github.com/user-attachments/assets/68d5e1e5-793d-40f3-a5bd-6a6dd0888595" />

##### Failed login

<img width="607" height="464" alt="image" src="https://github.com/user-attachments/assets/018580c1-99a4-41d8-9325-f793fdb95f14" />

(a bit unsightly but that is a known issue)

<img width="459" height="249" alt="image" src="https://github.com/user-attachments/assets/ac64c7f0-3840-4100-9a74-8707e43a208d" />


### Release Notes

#### New Features

- Add preview support for Snowflake Cortex language model provider with OAuth managed credentials integration

#### Bug Fixes

- N/A

### QA Notes

@:assistant

Added some basic tests in extensions/positron-assistant/src/test/snowflake.test.ts
- run manually with `npm run test-extension -- -l positron-assistant`

See product doc for additional background: https://docs.google.com/document/d/11n-snvHuyRvmzCpWFkR1btFk57ajBfkHvH78XYoq3KI

See current testing steps: https://docs.google.com/document/d/1fijhTQvgI1xUfJme_gMZtK3jOiVR6X_w7dtYsuZI-H8/edit?tab=t.e2zl4uys2eic#heading=h.3to4o6koos26

**Prerequisites**: 
- Positron Assistant is enabled in settings
- `snowflake-cortex` is added as an enabledProvider in settings
- One of the supported auth methods is used to authenticate to Snowflake in Positron

**Testing Steps**:
1. Navigate to "Configure Language Model Providers"
2. Verify "Snowflake Cortex" appears as a provider option
3. Select Snowflake Cortex - should show "OAuth (Managed)" authentication method or indicate the API Key is being used
4. Test chat functionality with available models:
   - Claude Sonnet 4
   - Claude Sonnet 4.5
   - Claude Haiku 4.5
   - GPT-5
   - GPT-4.1
5. Verify tool calling works, noting that the getPlot tool is not working for non-Anthropic providers
6. Test error handling by temporarily breaking credentials (should show helpful error messages)

sharon-wang committed:
add basic Snowflake Cortex language model support

auth with api key and base url

sharon-wang committed:
update model definitions

sharon-wang committed:
initial snowflake auth handling using autoconfigure

sharon-wang committed:
use placeholder input schema for snowflake

tools with input schema like

```
{
	type: 'object',
	properties: { },
	required: []
};
```

are not accepted by Snowflake, such as our get plot tool which takes no input

To get around that, we specify a placeholder input value for tools that don't require any input

sharon-wang committed:
document inline which snowflake models are erroring

sharon-wang committed:
fix ai sdk plot tool result message conversion

sharon-wang committed:
pass snowflake token through

sharon-wang committed:
move custom fetch to OpenAI provider

replacing `max_tokens` with `max_output_tokens` is relevant for all providers with OpenAI models, not just snowflake

sharon-wang committed:
create OpenAI-compatible custom fetch to handle model errors

add some workarounds for model errors related to our usage of the AI SDK v4, as well as a Snowflake Cortex-specific issue with certain models, where the role field in the response is sometimes an empty string instead of "assistant".

sharon-wang committed:
remove claude opus 4 from snowflake models

not listed at https://docs.snowflake.com/en/user-guide/snowflake-cortex/aisql#regional-availability, so seems to not be available

sharon-wang committed:
unify snowflake cred extraction with pwb generic handling

sharon-wang committed:
add auth error handling to bail early from connection check

sharon-wang committed:
make autoconfigured auth more prominent

also hide the radio buttons when autoconfigured auth is in use

sharon-wang committed:
re-remove model (accidentally re-introduced in rebase)

sharon-wang committed:
add snowflake-specific auth error message handling

sharon-wang committed:
add unit tests for Snowflake provider functionality

sharon-wang committed:
ensure base url is set properly when autoconfiguring Snowflake

timtmok commented on src/vs/workbench/contrib/positronAssistant/browser/icons/snowflake.tsx:
```diff
@@ -0,0 +1,28 @@
+/*---------------------------------------------------------------------------------------------
```
Can we also add the icon to https://github.com/posit-dev/positron/blob/d49b1e96dfbb289f3982c5f861b75e442a1484dc/src/vs/workbench/contrib/chat/browser/modelPicker/providerIcons.ts so it can show up in the model selector?

sharon-wang committed:
fix openai lump of coal and add Snowflake provider icon to model switcher

also needed to update whitespace indentation to get past precommit hygiene check "Bad whitespace indentation" for each of the SVGs

sharon-wang committed:
move snowflake error handling to overridden parseProviderError method

sharon-wang commented:
Thanks Tim! I've implemented your suggestions. While I was adding the snowflake provider icon, I fixed up the lump of coal issue with the openai icon too.

In my testing, the getPlot tool actually still isn't working -- perhaps this can be a follow-up as that's a general problem with OpenAI models and not specific to Snowflake: https://github.com/posit-dev/positron/issues/10759. If that sounds good, I'll remove the handling I introduced in this PR for getPlot which doesn't reliably work.

timtmok reviewed (dismissed):
LGTM

I agree that we can fix the issues with `getPlot` at a later date

sharon-wang committed:
Revert "fix ai sdk plot tool result message conversion"

This reverts commit 8d93cdb2f315765316c39d2dfe660a11d8e06ca5.

I thought this was working, but the model is not reliably reading the base64 image data. It was providing a "description" of the plot based on the session data / code run, rather than the actual image.

sharon-wang committed:
remove custom snowflake input schema handling, as it turns out, it wasn't making a difference for the getplot tool
